### ISSUE TITLE
`logs.restApi: true` fails to deploy

### ISSUE THREAD

**Issue by mikybars** [serverless/serverless#12637]
### Issue description

Whenever I set the `logs.restApi` flag to `true` I get this funny error message 

### Context

(node:10793) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
Serverless ϟ Support

This generates a report from your last Serverless Framework command (including any errors) to use for Github Issues, debugging w/ AI, or creating a support ticket w/ Serverless Inc.

Summary Report -----------------

# Service Overview
- Serverless Framework Version: 4.1.6
- Service Config File: serverless.yml
- Service Name: url-shortener
- Service Runtime: python3.11
- Service Stage: pro
- Service Region: eu-west-1

## Service Path
/Users/miguel.perez/dev/mikybars/miguelito-api

## Command
deploy --stage pro,--param restApiLogsEnabled=true

## Error Message
An error occurred: CustomApiGatewayAccountCloudWatchRole - Received response status [FAILED] from custom resource. Message returned: The role with name serverlessApiGatewayCloudWatchRole cannot be found. See details in CloudWatch Log: 2024/06/25/[$LATEST]3bc42a543b7e471387d64887f7c765b0 (RequestId: ************************************).

## Error Stacktrace
```
at file:///Users/miguel.perez/.serverless/releases/4.1.6/package/dist/sf-core.js:835:830
at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
at async AwsDeploy.update (file:///Users/miguel.perez/.serverless/releases/4.1.6/package/dist/sf-core.js:835:25478)
```

## Error Code
AWS_CLOUD_FORMATION_UPDATE_STACK_INTERNAL_CUSTOM_CREATE_FAILED

## Service Config
```
service: url-shortener
frameworkVersion: "4"
useDotenv: true
configValidationMode: error
plugins:
  - serverless-domain-manager
provider:
  name: aws
  runtime: python3.11
  region: eu-west-1
  stage: dev
  logs:
    restApi: ${param:restApiLogsEnabled}
  environment:
    MAINTAINER_EMAIL: "${env:MAINTAINER}"
    BUCKET_NAME: !Ref RedirectBucket
    LINKS_TABLE: !Ref LinksTable
    USERS_TABLE: !Ref UsersTable
    UNIQUES_TABLE: !Ref UniquesTable
    COUNTERS_TABLE: !Ref CountersTable
  apiGateway:
    shouldStartNameWithService: true
    request:
      schemas:
        create-link-model:
          name: CreateLink
          schema: ${file(models/create-link.schema.json)}
        update-link-model:
          name: UpdateLink
          schema: ${file(models/update-link.schema.json)}
        update-user-model:
          name: UpdateUser
          schema: ${file(models/update-user.schema.json)}
  iam:
    role:
      statements:
        - Effect: "Allow"
          Action:
            - ses:SendEmail
          Resource: "arn:aws:ses:*:*:identity/${env:MAINTAINER}"
        - Effect: "Allow"
          Action:
            - s3:ListBucket
          Resource: !GetAtt RedirectBucket.Arn
        - Effect: "Allow"
          Action:
            - s3:GetObject
            - s3:PutObject
            - s3:PutObjectAcl
            - s3:DeleteObject
          Resource: !Join ["/", [!GetAtt RedirectBucket.Arn, "*"]]
        - Effect: "Allow"
          Action:
            - dynamodb:UpdateItem
          Resource:
            - !GetAtt CountersTable.Arn
            - !GetAtt LinksTable.Arn
            - !GetAtt UsersTable.Arn
        - Effect: "Allow"
          Action:
            - dynamodb:PutItem
          Resource:
            - !GetAtt CountersTable.Arn
            - !GetAtt LinksTable.Arn
            - !GetAtt UniquesTable.Arn
            - !GetAtt UsersTable.Arn
        - Effect: "Allow"
          Action:
            - dynamodb:DeleteItem
          Resource:
            - !GetAtt LinksTable.Arn
            - !GetAtt UniquesTable.Arn
            - !GetAtt UsersTable.Arn
        - Effect: "Allow"
          Action:
            - dynamodb:Scan
            - dynamodb:GetItem
          Resource:
            - !GetAtt LinksTable.Arn
            - !GetAtt UniquesTable.Arn
            - !GetAtt UsersTable.Arn
        - Effect: "Allow"
          Action:
            - dynamodb:Query
            - dynamodb:BatchWriteItem
          Resource:
            - !GetAtt LinksTable.Arn
            - !GetAtt UniquesTable.Arn
        - Effect: "Allow"
          Action:
            - dynamodb:Query
          Resource:
            - !Join ["/", [!GetAtt LinksTable.Arn, "index", "User-index"]]
            - !Join ["/", [!GetAtt UsersTable.Arn, "index", "Email-index"]]
resources:
  - Conditions:
      IsPro: !Equals ["${sls:stage}", pro]
      IsDev: !Not [Condition: IsPro]
  - ${file(resources/db.cf.yml)}
  - ${file(resources/users.cf.yml)}
  - ${file(resources/web.cf.yml)}
stages:
  default:
    params:
      userPoolName: ${self:custom.rootDomain}-users-${sls:stage}
  dev:
    params:
      apiPath: dev
      corsOrigin: "http://localhost:8080"
      restApiLogsEnabled: false
      accessTokenValidityInMins: 60
      idTokenValidityInMins: 60
      refreshTokenValidityInMins: 600
      redirectBucketName: dev.${self:custom.rootDomain}
  pro:
    params:
      apiPath: v1
      corsOrigin: ${self:custom.rootDomain}
      restApiLogsEnabled: false
      accessTokenValidityInMins: 5
      idTokenValidityInMins: 5
      refreshTokenValidityInMins: 60
      redirectBucketName: ${self:custom.rootDomain}
custom:
  rootDomain: ${env:DOMAIN_NAME}
  certificateArn: ${env:CERTIFICATE_ARN}
  customDomain:
    autoDomain: true
    domainName: api.${env:DOMAIN_NAME}
    certificateArn: ${env:CERTIFICATE_ARN}
    basePath: ${param:apiPath}
    createRoute53Record: true
  authMapping: |-
    "auth": {
      "user": "$context.authorizer.claims.email"
    }
package:
  patterns:
    - "!**/*"
    - src/user/*.py
    - src/link/*.py
    - src/admin/*.py
    - src/common/*.py
functions:
  validateUserSignUp:
    handler: src/user/api.validate_user_signup
    description: Denies access to non activated user accounts while sending request-to-join emails the first time
    events:
      - cognitoUserPool:
          pool: ${param:userPoolName}
          trigger: PreSignUp
          existing: true
  updateUser:
    handler: src/user/api.update_user
    description: Updates a user details such as role or status (admin only)
    events:
      - http:
          path: /users/{userId}
          integration: lambda
          method: patch
          request:
            schemas:
              application/json: update-user-model
            template:
              application/x-www-form-urlencoded: null
              application/json: |
                {
                  "data": $input.json("$"),
                  "user_id": "$input.params().path.get('userId')",
                  ${self:custom.authMapping}
                }
            passThrough: NEVER
          response:
            statusCodes:
              200: &okResponse
                pattern: ""
              400: &userError
                pattern: "(ValidationError|ApplicationError): .*"
                template: |
                  #set ($msg = $input.path("$.errorMessage"))
                  #set ($index = $msg.indexOf(':') + 2)
                  { "message": "$msg.substring($index)" }
              403: &authError
                <<: *userError
                pattern: "AuthorizationError: .*"
              404: &notFoundError
                <<: *userError
                pattern: "NotFoundError: .*"
              500: &unexpectedError
                pattern: '.*\berror\b.*'
                template: '{ "message": "something went wrong" }'
          authorizer: &auth
            name: CognitoAuthorizer
            type: COGNITO_USER_POOLS
            arn:
              Fn::GetAtt:
                - CognitoUserPool
                - Arn
  listUsers:
    handler: src/user/api.list_users
    description: Lists all the users (admin only)
    events:
      - http:
          path: /users
          integration: lambda
          method: get
          request:
            template:
              application/x-www-form-urlencoded: null
              application/json: |
                {
                  ${self:custom.authMapping}
                }
            passThrough: NEVER
          response:
            statusCodes:
              200: *okResponse
              403: *authError
              500: *unexpectedError
          authorizer: *auth
  deleteUser:
    handler: src/user/api.delete_user
    description: Deletes a user (admin only)
    events:
      - http:
          path: /users/{userId}
          integration: lambda
          method: delete
          request:
            template:
              application/x-www-form-urlencoded: null
              application/json: |
                {
                  "user_id": "$input.params().path.get('userId')",
                  ${self:custom.authMapping}
                }
            passThrough: NEVER
          response:
            statusCodes:
              200: *okResponse
              403: *authError
              404: *notFoundError
              500: *unexpectedError
          authorizer: *auth
  apiInfo:
    handler: src/link/api.get_info
    description: Returns relevant information about the app such as the base URL
    events:
      - http:
          path: /info
          integration: lambda
          cors: true
          method: get
          response:
            statusCodes:
              200:
                pattern: ""
  createLink:
    handler: src/link/api.create_link
    description: Creates a new link pointing to the URL provided
    events:
      - http:
          path: /links
          integration: lambda
          method: post
          cors:
            origin: "${param:corsOrigin}"
          request:
            schemas:
              application/json: create-link-model
            template:
              application/x-www-form-urlencoded: null
              application/json: |
                {
                  "data": $input.json("$"),
                  ${self:custom.authMapping}
                }
            passThrough: NEVER
          response:
            statusCodes:
              201: *okResponse
              400: *userError
              403: *authError
              500: *unexpectedError
          authorizer: *auth
  listLinks:
    handler: src/link/api.list_links
    description: Lists all the links owned by the current user
    events:
      - http:
          path: /links
          integration: lambda
          method: get
          cors:
            origin: "${param:corsOrigin}"
          request:
            template:
              application/x-www-form-urlencoded: null
              application/json: |
                {
                  ${self:custom.authMapping}
                }
          response:
            statusCodes:
              200: *okResponse
              403: *authError
              500: *unexpectedError
          authorizer: *auth
  listLinksAsAdmin:
    handler: src/admin/api.list_all_links
    description: Lists all the links
    events:
      - http:
          path: /admin/links
          integration: lambda
          method: get
          request:
            template:
              application/x-www-form-urlencoded: null
              application/json: |
                {
                  ${self:custom.authMapping}
                }
          response:
            statusCodes:
              200: *okResponse
              403: *authError
              500: *unexpectedError
          authorizer: *auth
  deleteLink:
    handler: src/link/api.delete_link
    description: Deletes a link given its backhalf only if owned by the current user
    events:
      - http:
          path: /links/{backhalf}
          integration: lambda
          method: delete
          cors:
            origin: "${param:corsOrigin}"
          request:
            template:
              application/x-www-form-urlencoded: null
              application/json: |
                {
                  "backhalf": "$input.params().path.get('backhalf')",
                  ${self:custom.authMapping}
                }
          response:
            statusCodes:
              204: *okResponse
              403: *authError
              500: *unexpectedError
          authorizer: *auth
  deleteLinkAsAdmin:
    handler: src/admin/api.delete_link
    description: Deletes a link given its backhalf
    events:
      - http:
          path: /admin/links/{backhalf}
          integration: lambda
          method: delete
          request:
            template:
              application/x-www-form-urlencoded: null
              application/json: |
                {
                  "backhalf": "$input.params().path.get('backhalf')",
                  ${self:custom.authMapping}
                }
          response:
            statusCodes:
              204: *okResponse
              403: *authError
              404: *notFoundError
              500: *unexpectedError
          authorizer: *auth
  deleteAllLinks:
    handler: src/admin/api.delete_all_links
    description: Deletes all the links
    events:
      - http:
          path: /admin/links
          integration: lambda
          method: delete
          cors:
            origin: "${param:corsOrigin}"
          request:
            template:
              application/x-www-form-urlencoded: null
              application/json: |
                {
                  ${self:custom.authMapping}
                }
          response:
            statusCodes:
              204: *okResponse
              403: *authError
              500: *unexpectedError
          authorizer: *auth
  updateLink:
    handler: src/link/api.update_link
    description: Updates the link identified by its backhalf only if owned by the current user
    events:
      - http:
          path: /links/{backhalf}
          integration: lambda
          method: patch
          cors:
            origin: "${param:corsOrigin}"
          request:
            schemas:
              application/json: update-link-model
            template:
              application/x-www-form-urlencoded: null
              application/json: |
                {
                  "data": $input.json("$"),
                  "backhalf": "$input.params().path.get('backhalf')",
                  ${self:custom.authMapping}
                }
            passThrough: NEVER
          response:
            statusCodes:
              200: *okResponse
              400: *userError
              403: *authError
              500: *unexpectedError
          authorizer: *auth
  updateLinkAsAdmin:
    handler: src/admin/api.update_link
    description: Updates the link identified by its backhalf
    events:
      - http:
          path: /admin/links/{backhalf}
          integration: lambda
          method: patch
          cors:
            origin: "${param:corsOrigin}"
          request:
            schemas:
              application/json: update-link-model
            template:
              application/x-www-form-urlencoded: null
              application/json: |
                {
                  "data": $input.json("$"),
                  "backhalf": "$input.params().path.get('backhalf')",
                  ${self:custom.authMapping}
                }
            passThrough: NEVER
          response:
            statusCodes:
              200: *okResponse
              400: *userError
              404: *notFoundError
              403: *authError
              500: *unexpectedError
          authorizer: *auth
```

Above is a summary report of your last Serverless Framework command.

Use it to debug with AI (e.g. ChatGPT) or report a new Github Issue: https://slss.io/issue

**Comment by eahefnawy** [comment]
Thanks for reporting @mikybars ... We'll look into this.

**Comment by eahefnawy** [comment]
@mikybars I am unable to reproduce this at all. What makes you think this it's related to rest api logs? The error message seem to indicate that a role is missing?

**Comment by mikybars** [comment]
Well, that's easy. I make two deployments: one with the flag equals to `false` and the other one with the flag equals to `true` (that's being the only difference between the two). The first one works 👍🏾 and the second one 👎🏾

**Comment by mikybars** [comment]
For debugging purposes, I attach the diff between the packages generated with both settings:

```diff
--- cloudformation-template-update-stack-false.json	2024-06-28 11:04:51
+++ cloudformation-template-update-stack-true.json	2024-06-28 11:02:36
@@ -3391,8 +3391,26 @@
             ]
           ]
         }
+      }
+    },
+    "ApiGatewayLogGroup": {
+      "Type": "AWS::Logs::LogGroup",
+      "Properties": {
+        "LogGroupName": "/aws/api-gateway/url-shortener-pro"
       }
     },
+    "CustomApiGatewayAccountCloudWatchRole": {
+      "Type": "Custom::ApiGatewayAccountRole",
+      "Version": "1.0",
+      "Properties": {
+        "ServiceToken": {
+          "Fn::GetAtt": [
+            "CustomDashresourceDashapigwDashcwDashroleLambdaFunction",
+            "Arn"
+          ]
+        }
+      }
+    },
     "IamRoleCustomResourcesLambdaExecution": {
       "Type": "AWS::IAM::Role",
       "Properties": {
@@ -3429,6 +3447,50 @@
               "Statement": [
                 {
                   "Effect": "Allow",
+                  "Resource": {
+                    "Fn::Join": [
+                      ":",
+                      [
+                        "arn",
+                        {
+                          "Ref": "AWS::Partition"
+                        },
+                        "iam:",
+                        {
+                          "Ref": "AWS::AccountId"
+                        },
+                        "role/*"
+                      ]
+                    ]
+                  },
+                  "Action": [
+                    "iam:AttachRolePolicy",
+                    "iam:CreateRole",
+                    "iam:ListAttachedRolePolicies",
+                    "iam:PassRole"
+                  ]
+                },
+                {
+                  "Effect": "Allow",
+                  "Resource": {
+                    "Fn::Join": [
+                      ":",
+                      [
+                        "arn",
+                        {
+                          "Ref": "AWS::Partition"
+                        },
+                        "apigateway:*::/account"
+                      ]
+                    ]
+                  },
+                  "Action": [
+                    "apigateway:GET",
+                    "apigateway:PATCH"
+                  ]
+                },
+                {
+                  "Effect": "Allow",
                   "Resource": "*",
                   "Action": [
                     "cognito-idp:ListUserPools",
@@ -3461,6 +3523,31 @@
         ]
       }
     },
+    "CustomDashresourceDashapigwDashcwDashroleLambdaFunction": {
+      "Type": "AWS::Lambda::Function",
+      "Properties": {
+        "Code": {
+          "S3Bucket": {
+            "Ref": "ServerlessDeploymentBucket"
+          },
+          "S3Key": "serverless/url-shortener/pro/1719565193467-2024-06-28T08:59:53.467Z/custom-resources.zip"
+        },
+        "FunctionName": "url-shortener-pro-custom-resource-apigw-cw-role",
+        "Handler": "api-gateway-cloud-watch-role/handler.handler",
+        "MemorySize": 1024,
+        "Runtime": "nodejs20.x",
+        "Timeout": 180,
+        "Role": {
+          "Fn::GetAtt": [
+            "IamRoleCustomResourcesLambdaExecution",
+            "Arn"
+          ]
+        }
+      },
+      "DependsOn": [
+        "IamRoleCustomResourcesLambdaExecution"
+      ]
+    },
     "ValidateUserSignUpCustomCognitoUserPool1": {
       "Type": "Custom::CognitoUserPool",
       "Version": "1.0",
```

**Comment by eahefnawy** [comment]
Thanks @mikybars ... This is **really** helpful! I think I know what the issue _might_ be. Let me try something on my end. Meanwhile, to get around this error, you can try to create the API Gateway role yourself, and reference it in `serverless.yml`:

```yml
# serverless.yml
provider:
  logs:
    restApi:
      role: arn:aws:iam::123456:role
```

The role should look something like this:

```json
{
    "Version": "2012-10-17",
    "Statement": [
      {
        "Effect": "Allow",
        "Principal": {
          "ervice": ["apigateway.amazonaws.com"]
        },
        "Action": ["sts:AssumeRole"]
      }
    ]
}
```
Then attach the following policy to it:
```
arn:aws:iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs
```

**Comment by blongv** [comment]
I can confirm I just ran into the same issue.  Creating the role manually and assigning it works perfectly.
```
custom:
  cloudwatchRoles:
    nonprod: arn:aws:iam::12345:role/tempServerlessApiGatewayCloudWatchRole
    prod: arn:aws:iam::67890:role/tempServerlessApiGatewayCloudWatchRole

provider:
  logs:
    restApi:
      role: ${self:custom.cloudwatchRoles.${sls:stage}}
```

**Comment by mikybars** [comment]
I can confirm that the workaround is working for me too 👍🏾

**Comment by eahefnawy** [comment]
Alright, this was pretty tricky to track down, but I was finally able to reproduce and fix this issue in https://github.com/serverless/serverless/pull/12655. The fix will be included in the next release 🚀 

@mikybars Thanks a lot for the detailed issue report and for your engagement! 🙌

**Comment by eahefnawy** [comment]
This is now fixed & released in [v4.1.8](https://github.com/serverless/serverless/releases/tag/v4.1.8).

**Comment by Trezcool** [comment]
hi @eahefnawy what about v3?
we are experiencing the same on v3
